### ISSUE TITLE
Weave proxy losing docker event updates

### ISSUE THREAD

**Issue by bboreham** [weaveworks/weave#2016]
As mentioned at https://github.com/weaveworks/weave/issues/1775#issuecomment-190734573, [these logs](https://gist.github.com/errordeveloper/8058d847572e1786bce0) show a proxy failing to receive Docker events, and subsequent investigation shows the goroutine for `fsouza/go-dockerclient` missing, presumed exited because of unexpected error.

**Comment by errordeveloper** [comment]
```
INFO: 2016/03/14 12:17:05.759665 weave proxy awh1
INFO: 2016/03/14 12:17:05.759752 Command line arguments: --rewrite-inspect --tlsverify --tlscacert /home/weave/tls/ca.pem --tlscert /home/weave/tls/cert.pem --tlskey /home/weave/tls/key.pem -H unix:///var/run/weave/weave.sock
INFO: 2016/03/14 12:17:05.930790 Docker API on unix:///var/run/docker.sock: &[Arch=amd64 KernelVersion=4.1.18-boot2docker BuildTime=2016-02-22T22:37:33.778002647+00:00 Version=1.10.2 ApiVersion=1.22 GitCommit=c3959b1 GoVersion=go1.5.3 Os=linux]
2016/03/14 12:17:05 [go-dockerclient] Adding event listener
INFO: 2016/03/14 12:17:05.958362 proxy listening on unix:///var/run/weave/weave.sock
2016/03/14 12:17:05 [go-dockerclient] entering monitorEvents
2016/03/14 12:17:05 [go-dockerclient] eventHijack new request &{GET /events HTTP/1.1 1 1 map[] <nil> 0 [] false  map[] map[] <nil> map[]   <nil>}
2016/03/14 12:17:05 [go-dockerclient] eventHijack waiting for event
2016/03/14 12:17:05 [go-dockerclient] eventHijack event decode error EOF
2016/03/14 12:17:05 [go-dockerclient] eventHijack IO EOF EOF
INFO: 2016/03/14 12:17:07.852286 POST /v1.22/containers/create?name=toolbox-at-node-1
INFO: 2016/03/14 12:17:07.852453 Creating container with WEAVE_CIDR ""
INFO: 2016/03/14 12:17:07.928718 POST /v1.22/containers/19880eb45403fa5c184d71f3d26f283949e10ff19b60d6045a9c9c1cbb0a2b72/start
```

**Comment by awh** [comment]
For the record that debug output is coming from https://github.com/awh/go-dockerclient/commit/8b8ff40d8a9b40cbeddb1a0c4a899119767835ac

**Comment by awh** [comment]
@errordeveloper new build `awh2`

I have added event request body logging (https://github.com/awh/go-dockerclient/commit/ba126a4307c48ebaadfc079336c8a9f61aeb848a) - you should now see logging such as
```
2016/03/14 14:31:10 [go-dockerclient] eventHijack waiting for event
2016/03/14 14:31:10 [go-dockerclient] [{"Type":"network","Action":"disconnect","Actor":{"ID":"389ce74365bfc20932fbc743727732e5f580a46f8b43f287548c7f62caef3e3b","Attributes":{"container":"ce08ea0a583ec71fbed3c7e03cc6cb19539d4cba49e80c31b86ecef725d9bc35","name":"host","type":"host"}},"time":1457965870,"timeNano":1457965870029973006}
{"status":"destroy","id":"ce08ea0a583ec71fbed3c7e03cc6cb19539d4cba49e80c31b86ecef725d9bc35","from":"weaveworks/weaveexec:awh2","Type":"container","Action":"destroy","Actor":{"ID":"ce08ea0a583ec71fbed3c7e03cc6cb19539d4cba49e80c31b86ecef725d9bc35","Attributes":{"image":"weaveworks/weaveexec:awh2","name":"drunk_perlman","works.weave.role":"system"}},"time":1457965870,"timeNano":1457965870031932800}
] 693 <nil>
2016/03/14 14:31:10 [go-dockerclient] decoded event {   1457965870}
```

**Comment by errordeveloper** [comment]
```
INFO: 2016/03/14 14:41:26.021134 weave proxy awh2
INFO: 2016/03/14 14:41:26.021248 Command line arguments: --rewrite-inspect --tlsverify --tlscacert /home/weave/tls/ca.pem --tlscert /home/weave/tls/cert.pem --tlskey /home/weave/tls/key.pem -H unix:///var/run/weave/weave.sock
INFO: 2016/03/14 14:41:26.153561 Docker API on unix:///var/run/docker.sock: &[GoVersion=go1.5.3 Os=linux Arch=amd64 KernelVersion=4.1.18-boot2docker BuildTime=2016-02-22T22:37:33.778002647+00:00 Version=1.10.2 ApiVersion=1.22 GitCommit=c3959b1]
2016/03/14 14:41:26 [go-dockerclient] Adding event listener
INFO: 2016/03/14 14:41:26.191774 proxy listening on unix:///var/run/weave/weave.sock
2016/03/14 14:41:26 [go-dockerclient] entering monitorEvents
2016/03/14 14:41:26 [go-dockerclient] eventHijack new request &{GET /events HTTP/1.1 1 1 map[] <nil> 0 [] false  map[] map[] <nil> map[]   <nil>}
2016/03/14 14:41:26 [go-dockerclient] eventHijack waiting for event
2016/03/14 14:41:26 [go-dockerclient] [] 0 EOF
2016/03/14 14:41:26 [go-dockerclient] eventHijack event decode error EOF
2016/03/14 14:41:26 [go-dockerclient] eventHijack IO EOF EOF
INFO: 2016/03/14 14:41:27.982127 POST /v1.22/containers/create?name=toolbox-at-node-2
INFO: 2016/03/14 14:41:27.982280 Creating container with WEAVE_CIDR ""
INFO: 2016/03/14 14:41:28.064941 POST /v1.22/containers/e7832dc85908d82858a3f73d3c9ddd24ebd76e1158fc046d5c6c469a6921c054/start
```

**Comment by awh** [comment]
Interesting, so it's getting an EOF immediately on trying to decode the first event.

**Comment by awh** [comment]
We can mitigate this ourselves by having `AddObserver` in `common/docker/client.go` resubscribe with a back-off if the channel closes.

**Comment by rade** [comment]
Fixed by #2049.
